File: computing/interview/design-patterns/creational/builder/csharp/BuilderDemo.cs
using System;
using System.Collections.Generic;

namespace Interview.DesignPatterns.Creational.Builder
{
    public sealed class Meal
    {
        private readonly List<string> _items;
        private readonly int _calories;

        public Meal(IEnumerable<string> items, int calories)
        {
            _items = new List<string>(items);
            _calories = calories;
        }

        public void Describe()
        {
            Console.WriteLine($"Meal: {string.Join(" ", _items)} ({_calories} kcal)");
        }
    }

    public sealed class MealBuilder
    {
        private readonly List<string> _items = new();
        private int _calories;

        public MealBuilder AddMain(string name, int calories)
        {
            _items.Add(name);
            _calories += calories;
            return this;
        }

        public MealBuilder AddSide(string name, int calories)
        {
            _items.Add(name);
            _calories += calories;
            return this;
        }

        public MealBuilder AddDrink(string name, int calories)
        {
            _items.Add(name);
            _calories += calories;
            return this;
        }

        public Meal Build()
        {
            var meal = new Meal(_items, _calories);
            _items.Clear();
            _calories = 0;
            return meal;
        }
    }

    public sealed class MealDirector
    {
        public Meal CreateHighProtein(MealBuilder builder) => builder
            .AddMain("Grilled chicken", 400)
            .AddSide("Quinoa", 180)
            .AddDrink("Protein shake", 220)
            .Build();

        public Meal CreateVegetarian(MealBuilder builder) => builder
            .AddMain("Tofu stir-fry", 320)
            .AddSide("Salad", 90)
            .AddDrink("Green tea", 0)
            .Build();
    }

    public static class Demo
    {
        public static void Main()
        {
            var builder = new MealBuilder();
            var director = new MealDirector();

            director.CreateHighProtein(builder).Describe();
            director.CreateVegetarian(builder).Describe();
        }
    }
}

File: computing/interview/design-patterns/creational/builder/java/BuilderDemo.java
package interview.designpatterns.creational.builder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

final class Meal {
    private final List<String> items;
    private final int calories;

    Meal(List<String> items, int calories) {
        this.items = List.copyOf(items);
        this.calories = calories;
    }

    void describe() {
        System.out.printf("Meal: %s (%d kcal)%n", String.join(" ", items), calories);
    }
}

final class MealBuilder {
    private final List<String> items = new ArrayList<>();
    private int calories;

    MealBuilder addMain(String name, int cals) {
        items.add(name);
        calories += cals;
        return this;
    }

    MealBuilder addSide(String name, int cals) {
        items.add(name);
        calories += cals;
        return this;
    }

    MealBuilder addDrink(String name, int cals) {
        items.add(name);
        calories += cals;
        return this;
    }

    Meal build() {
        Meal meal = new Meal(items, calories);
        items.clear();
        calories = 0;
        return meal;
    }
}

final class MealDirector {
    Meal createHighProtein(MealBuilder builder) {
        return builder
            .addMain("Grilled chicken", 400)
            .addSide("Quinoa", 180)
            .addDrink("Protein shake", 220)
            .build();
    }

    Meal createVegetarian(MealBuilder builder) {
        return builder
            .addMain("Tofu stir-fry", 320)
            .addSide("Salad", 90)
            .addDrink("Green tea", 0)
            .build();
    }
}

public final class BuilderDemo {
    private BuilderDemo() {}

    public static void main(String[] args) {
        MealBuilder builder = new MealBuilder();
        MealDirector director = new MealDirector();
        director.createHighProtein(builder).describe();
        director.createVegetarian(builder).describe();
    }
}
